collect_app/src/main/java/org/odk/collect/android/formmanagement/FormsUpdater.kt

```diff
@@ -0,0 +1,182 @@
+                    .collect(Collectors.toList())
+            if (updatedForms.isNotEmpty()) {
+                if (sandbox.generalSettings.getBoolean(GeneralKeys.KEY_AUTOMATIC_UPDATE)) {
+                    val formUpdateDownloader = FormUpdateDownloader()
+                    val results = formUpdateDownloader.downloadUpdates(
+                        updatedForms,
+                        sandbox.formsLock,
+                        formDownloader,
+                        TranslationHandler.getString(context, R.string.success),
+                        TranslationHandler.getString(context, R.string.failure)
+                    )
+
+                    notifier.onUpdatesDownloaded(results)
+                } else {
+                    notifier.onUpdatesAvailable(updatedForms)
+                }
+            }
+
+            context.contentResolver.notifyChange(FormsProviderAPI.CONTENT_URI, null)
+        } catch (_: FormSourceException) {
+            // Ignored
+        }
+    }
+
+    /**
+     * Downloads new forms, updates existing forms and deletes forms that are no longer part of
+     * the project's form list.
+     */
+    fun matchFormsWithServer(projectId: String): Boolean {
+        val sandbox = getProjectSandbox(projectId)
+
+        val diskFormsSynchronizer = diskFormsSynchronizer(sandbox)
+        val serverFormsDetailsFetcher = serverFormsDetailsFetcher(sandbox, diskFormsSynchronizer)
+        val formDownloader = formDownloader(sandbox, analytics)
+
+        val serverFormsSynchronizer = ServerFormsSynchronizer(
+            serverFormsDetailsFetcher,
+            sandbox.formsRepository,
+            sandbox.instancesRepository,
+            formDownloader
+        )
+
+        return sandbox.formsLock.withLock { acquiredLock ->
+            if (acquiredLock) {
+                syncStatusAppState.startSync()
+
+                val exception = try {
+                    serverFormsSynchronizer.synchronize()
+                    syncStatusAppState.finishSync(null)
+                    notifier.onSync(null)
+                    null
+                } catch (e: FormSourceException) {
+                    syncStatusAppState.finishSync(e)
+                    notifier.onSync(e)
+                    e
+                }
+
+                AnalyticsUtils.logMatchExactlyCompleted(analytics, exception)
+                exception == null
+            } else {
+                false
+            }
+        }
+    }
+
+    private fun getProjectSandbox(projectId: String) = ProjectSandbox(
+        projectId,
+        settingsProvider,
+        formsRepositoryProvider,
+        instancesRepositoryProvider,
+        storagePathProvider,
+        changeLockProvider,
+        formSourceProvider
+    )
+}
+
+/**
+ * Provides all the basic/building block dependencies needed when performing logic inside a
+ * project.
+ */
+private class ProjectSandbox(
```

Comment:
I could see this ending up being a convenience we share in a few places (I think we'll need it almost immediately for auto send for instance). This could also be implemented with a Dagger component (that we just build on the fly). But personally I think I'm happy to avoid Dagger in places it's not absolutely needed (i.e. where we have a constructor).